webapp/sources/rudder/rudder-web/src/main/scala/com/normation/plugins/RudderPluginJson.scala

```diff
@@ -0,0 +1,131 @@
+
+*
+*************************************************************************************
+*/
+
+package com.normation.plugins
+
+import java.nio.charset.StandardCharsets
+
+import better.files.File
+import org.joda.time.DateTime
+import zio._
+import zio.syntax._
+import com.normation.errors._
+import net.liftweb.json._
+import org.joda.time.format.ISODateTimeFormat
+
+/*
+ * This file is able to read information from /var/rudder/packages/index.json
+ * (package management file for plugins)
+ */
+
+// this is only used internally but liftweb can't hand it if not top level
+protected final case class JsonPluginFile(plugins: Map[String, JValue])
+
+// this is only used internally but liftweb can't hand it if not top level
+protected final case class JsonPluginRaw(
+    name          : String
+  , version       : String
+  , files         : List[String]
+  , `jar-files`   : List[String]
+  , `build-commit`: String
+  , `build-date`  : String
+)
+
+final case class JsonPluginDef(
+    name       : String
+  , version    : PluginVersion
+  , files      : List[String]
+  , jars       : List[String]
+  , buildCommit: String
+  , buildDate  : DateTime
+)
+
+/*
+ * The service in charge of reading the package file with plugin info.
+ * By convention, in Rudder, path = /var/rudder/packages/index.json
+ */
+class ReadPluginPackageInfo(path: String) {
+
+  val index = File(path)
+
+  def check: IOResult[Boolean] = {
+    IOResult.effect {
+      index.isRegularFile && index.isReadable
+    }
+  }
+
+  def readIndex(): IOResult[String] = {
+    IOResult.effect {
+      index.contentAsString(StandardCharsets.UTF_8)
+    }
+  }
+
+  def parseFile(json: String): IOResult[JsonPluginFile] = {
+    IOResult.effect {
+      implicit val formats = net.liftweb.json.DefaultFormats
+      JsonParser.parse(json).extract[JsonPluginFile]
+    }
+  }
+
+  def decodeOne(plugin: JValue): IOResult[JsonPluginDef] = {
+    implicit val formats = net.liftweb.json.DefaultFormats
+    for {
+      p       <- IOResult.effect(plugin.extract[JsonPluginRaw])
+      date    <- IOResult.effect(DateTime.parse(p.`build-date`, ISODateTimeFormat.dateTimeNoMillis()))
+      version <- PluginVersion.from(p.version).notOptional(s"Version for '${p.name}' is not a valid plugin version: ${p.version}")
+    } yield {
+      JsonPluginDef(p.name, version, p.files, p.`jar-files`, p.`build-commit`, date)
+    }
+  }
+
+  def decode(plugin: JsonPluginFile): IOResult[List[Either[RudderError, JsonPluginDef]]] = {
+    ZIO.foreach(plugin.plugins) { case (name, jvalue) =>
+      decodeOne(jvalue).mapError(err => Chained(s"Error when decoding plugin information for entry '${name}'", err)).either
+    }
+  }
+
+  def parseJson(json: String) = {
+    parseFile(json) >>= decode
+  }
+
+  def getInfo(): IOResult[List[Either[RudderError, JsonPluginDef]]] = {
+    IOResult.effect(index.exists).flatMap { exists =>
+      if(exists) check *> readIndex() >>= parseJson
```

Comment:
wow, i don't know this syntax

Reply:
`*>` means "exec what's on left and discard result on success", and `>>=` means `flatMap` (and it binds the result to the next method param)